dnf-json

```diff
@@ -200,64 +250,129 @@ class DnfJsonRequestHandler(BaseHTTPRequestHandler):
+                        self.response_success(solver.dump())
+                        log.info("dump success")
+                    elif command == "depsolve":
+                        self.response_success(
+                                solver.depsolve(
+                                    arguments["package-specs"],
+                                    arguments.get("exclude-specs", [])
+                                    )
                                 )
-                            )
-                    log.info("depsolve success")
-
-            except dnf.exceptions.MarkingErrors as e:
-                log.info("error install_specs")
-                self.response_with_dnf_error(
-                    "MarkingErrors",
-                    f"Error occurred when marking packages for installation: {e}"
-                )
-            except dnf.exceptions.DepsolveError as e:
-                log.info("error depsolve")
-                self.response_with_dnf_error(
-                    "DepsolveError",
-                    (
-                        "There was a problem depsolving "
-                        f"{arguments['package-specs']}: {e}"
+                        log.info("depsolve success")
+
+                except dnf.exceptions.MarkingErrors as e:
+                    log.info("error install_specs")
+                    self.response_with_dnf_error(
+                        "MarkingErrors",
+                        f"Error occurred when marking packages for installation: {e}"
                     )
-                )
-            except dnf.exceptions.Error as e:
-                self.response_with_dnf_error(
-                    type(e).__name__,
-                    f"Error occurred when setting up repo: {e}")
+                except dnf.exceptions.DepsolveError as e:
+                    log.info("error depsolve")
+                    self.response_with_dnf_error(
+                        "DepsolveError",
+                        (
+                            "There was a problem depsolving "
+                            f"{arguments['package-specs']}: {e}"
+                        )
+                    )
+                except dnf.exceptions.Error as e:
+                    self.response_with_dnf_error(
+                        type(e).__name__,
+                        f"Error occurred when setting up repo: {e}")
+        finally:
+            cache = None
+            with process_count.get_lock():
+                process_count.value -= 1
+                # Only the last process alive is making the cleaning.
+                # This is because we don't want to delete directories that might
+                # be used by a live dnf instance.
+                if process_count.value == 0:
+                    self.clean_cache()
+                else:
+                    # If the process is not the last, it sends it's used folders
+                    # to the last one.
+                    self.send_cache_folder_used()
```

Comment:
I think the code might be easier to understand if we always call this method. IOW I would put this line above `if process_count.value == 0:`. Note that `clean_cache` must be slightly adjusted (`self.cache_folders` can't be reused in this case which is IMHO good.)

Reply:
This could be done even without adjusting the `clean_cache` method. Because it will fetch all the caches send through the queue and eventually it's own. But yes, maybe cleaner without having the process sending itself its own stuff.
But yeah, I've been thinking about for some time now. The thing is there can be some kind of starvation where in the worst case scenario the cache is never cleaned. IOW (you taught me this one today :smile: ) if there's always a new process answering a request arriving, then we never reach `0` in the counter, then the `clean_cache` is never called.
What I want to do is the following:
* Having another process whose work is to receive name of cache folders, update the file on disk and clean stuff.
* Make this process clean the cache at fix intervals, let's say every 50 requests.
* Make the other processes hold when the cleaning occurs.

Reply:
I wonder if locking per a cache folder rather than per a whole cache can solve this issue. Then the cleaner process can just process unlocked directories and ignore locked ones (and it doesn't actually make sense to clean these because they are being used).

Reply:
Hmm, I wonder if we can just use `mtime`, or `atime` on these directories & per-directory locking to decide whether they can be deleted...

Reply:
We can't rely change the locking mechanism as it is handled by the `dnf` library. However we could have multiple cache directories allowing multiple requests to be answered at the same time. I've changer the whole thing to fit my previous plan. I just need to find where to declare that dnf-json know requires this lib: https://pypi.org/project/prwlock/